### DISCUSSION TITLE
Is it mandatory for servers to honor the defer/stream directive?

### DISCUSSION THREAD

**Discussion by robrichard** [graphql/defer-stream-wg#7] [Architectural Decision Records]
## Decision
The spec will state that servers SHOULD honor the `@defer` and `@stream` directives. 

## Consequences
* The ability to defer/stream parts of a response can have a potentially significant impact on application performance. Developers generally need clear, predictable control over their application's performance. It is highly recommended that the GraphQL server honor the @defer and @stream directives on each execution. However, the specification will allow advanced use-cases where the server can determine that it is more performant to not defer/stream. Therefore, GraphQL clients must be able to process a response that ignores the defer/stream directives. See discussion with josephsavona (https://github.com/graphql/graphql-spec/issues/704#issuecomment-613438677)
* This also applies to the initialCount argument on the @stream directive. Clients must be able to process a streamed response that contains a different number of initial list items than what was specified in the initialCount argument.
* We will avoid using the term "hint" to describe this behavior, since it is highly recommended that servers do not ignore defer/stream directives.
* Servers that do not support defer/stream must not ignore these directives. They should error by failing the [Directives Are Defined](https://spec.graphql.org/June2018/#sec-Directives-Are-Defined) validation rule.

Most importantly, leaving open the possibility that defer/stream can be ignored ensures that clients remain flexible. This allows for spec changes in the future that are non-breaking. If this ends up being the wrong decision, a new RFC can be brought to the working group to change SHOULD to MUST without fear of breaking clients. If we go with MUST now, it would not be possible to change to SHOULD or MAY without breaking clients.

## Status
- [x] defined in spec
- [x] implemented in graphql-js

**Comment by yaacovCR** [comment]
I think SHOULD could just be changed to MUST.

why does the server know better than the client what is more performant, and certainly will always take more bandwidth to send rather than defer.

perhaps an argument could be added ‘@defer unlessYouThinkNot=true’

i also don’t think we can say this is implemented in graphql js, maybe I missed it, is there a case where graphql js chooses not to honor the directive?

**Reply by maraisr** [reply]
Tend to agree, recently was in a chat about this as well => https://github.com/graphql/graphql-spec/pull/742#issuecomment-964861201

`@defer(optional: true)` has my vote. (`optional: false` default)

**Reply by robrichard** [reply]
We discussed with the WG and Facebook a while back and settled on the “SHOULD, but highly recommended” language.

Leaving open the possibility that defer/stream can be ignored ensures that clients remain flexible. This allows for spec changes in the future that are non-breaking, e.g SHOULD can always change to MUST but not vice versa. If we go with MUST now, it closes off those possibilities in the future, when we may be more informed after it is widely adopted.

Facebook also had some cases where their server does not honor defer/stream, but recommended that it should normally be honored. This is why we have the “advanced use-case” exception.

I don’t think it’s necessary for graphql-js to contain conditions where it does not honor defer/stream to be spec complaint. 

What we’re saying is all servers that support defer/stream SHOULD honor it, but if they have a good reason not to, they are still spec compliant.

**Reply by yaacovCR** [reply]
https://github.com/graphql/graphql-spec/pull/742#discussion_r747798161

**Reply by yaacovCR** [reply]
https://github.com/graphql/graphql-spec/pull/742#discussion_r747799488

**Reply by yaacovCR** [reply]
Echoing my last linked comment above, I would love to see some actual numbers in the scenario that the server determines more performant to not defer. What are the actual costs of deferring when you don’t need to?

Some hard numbers would probably convince me that this is worthwhile. Right now I just don’t see it, seems like added complexity for undefined gain.

**Reply by maraisr** [reply]
Is it okay to disagree? I'm far more fond on the idea and as a frontend developer, if I want to defer something, that it's being deferred. 😔 I'm okay for the spec to be less strict now, but not a "it's a hint" but more a "it's a hint" but highly encouraged to adhere. 

The hot chocolate folk 100% against this, and see it as a hint... and no more.

**Reply by robrichard** [reply]
We're intentionally using imprecise terms like "should" and "advanced use-cases" to be flexible enough to allow experimentation. I think once there is wide adoption, different approaches can be compared and best practices will become more apparent.

IMO hot chocolate's implementation fits the definition of "advanced use-cases" and could be a useful comparison point in the future.

Ultimately we want to avoid these two situations:

1. GraphQL clients require that the server honors all `@defer` & `@stream` directives. This is why it's SHOULD and not MUST, so clients have to be capable of handling ignored defer/stream to be spec compliant. If they didn't, it would limit the possibility for future optimization.
2. Developers choose not to use defer/stream and instead utilize other less performant query optimization techniques (e.g. multiple queries) because they can't rely on their server honoring the directives. We don't want to negate the usefulness of this feature.

I'm trying to avoid the word "hint" because I think that implies it's ok to arbitrarily ignore the directives. For example, a proxy that silently drops or rewrites responses to remove deferred payloads would not be ok.

**Reply by yaacovCR** [reply]
We could be precise and allow server flexibility by including an argument that specifically, explicitly allows server flexibility.

**Reply by maraisr** [reply]
Ya I'd definitely like that better, `@defer(optional: true)`; which leaves the server complete freedom to adhere to or not. And for moments where the client needs it to behave in a way, just optimal: false.

**Reply by yaacovCR** [reply]
> We're intentionally using imprecise terms like "should" and "advanced use-cases" to be flexible enough to allow experimentation. I think once there is wide adoption, different approaches can be compared and best practices will become more apparent.

I want to add that I appreciate the above point as a value, I just think that the current compromise in terms of specification language does not do enough to reflect the nuances of the underlying reasoning.

Taking a step back, my understanding of the goal of incremental delivery is something like the following: "overall most speedy result delivery (with client side suggestion of prioritization of the specified initial payload)."

According to the proposed edits, the `@defer` and `@stream` remain the client's best guess as to optimal performance, not a hint per se...but kind of a client-side (strongly-held?) belief? Servers are allowed to experiment to provide even better performance.

From the proposed edits:

> the specification allows advanced use-cases where the server can determine that it is more performant to not defer and/or stream

Are servers encouraged to experiment? Are they discouraged from experimenting? Is the specification...neutral? This is imprecise. It also gives off the impression of complexity, where in truth none need exist.

The advantage of the argument approach is that it makes clear that server experimentation is encouraged when allowed. If this suggestion is considered premature, and we still wish to allow servers to experiment without additional arguments, I think the specification should state so explicitly.

Specifically, I would suggest that we drop the simple use of SHOULD to describe how the server should approach the directives. We should instead use language something like the following:

```
- `if: Boolean` - When `true`, fragment *should* be deferred. When `false`, fragment will not be deferred and data will be included in the initial response. If omitted, defaults to `true`. 
+ `if: Boolean` - When `true`, fragment *should* be deferred, unless the server considers it more performant to not defer. When `false`, fragment will not be deferred and data will be included in the initial response. If omitted, defaults to `true`. 
```

The comment at the end should instead point out that servers are allowed to remain agnostic as to what is more performant, and simply respect the client's belief that incremental delivery is the way to go.

These are small changes, but at least accurately reflect where we currently are.

I also find the parallel between the client vs server-controlled nullability and client vs server controlled incremental delivery very interesting. The server has the best understanding of the underlying implementation, but the client has the best understanding of its needs, in both cases. I think we should do our best to align the client vs server modifers in the two cases, although I am not yet sure how this would work.

**Reply by yaacovCR** [reply]
@robrichard  any thoughts on slight adjustment in spec language above?

**Reply by robrichard** [reply]
We have almost this exact wording right below that section https://github.com/graphql/graphql-spec/pull/742/files#diff-30a69c5a5eded8e1aea52e53dad1181e6ec8f549ca2c50570b035153e2de1c43R2146
> the server can determine that it is more performant to not defer and/or stream

I'd prefer to keep it there where the full context is explained.

**Reply by yaacovCR** [reply]
Maybe then what we need is some language that more generally specifies that client side “directives” may indicate mandatory server behavior in some cases, but also may indicate only recommended behavior, corresponding to the must/should keywords as used in the spec. That should probably be a separate spec pr that comes before these edits. What do you think?

**Comment by n1ru4l** [comment]
I think the following questions are also related to this: https://github.com/graphql/graphql-over-http/issues/167

**Comment by mjmahone** [comment]
One reason why *should* is better for advanced use cases (maybe not a *great* reason but a reason):

If you have a server that composes many child servers (so each of those child servers' schemas compose to form the parent schema): some of those child servers may implement @defer/@stream, but if even one child server responsible for even a single schema type does not support it, you either need to walk the entire tree in the parent server in order to split up the already-stitched-together response tree (if all servers that support @defer/@stream *must* batch every time), or we need to allow ignoring the defer/stream directives when the server chooses.

**Reply by yaacovCR** [reply]
Supporting defer/stream in a gateway is something I am super interested in!

Let’s take defer and stream separately. For defer, my presumption is that if a child server does not support defer, one can best mimic the functionality by sending multiple queries and hoping the child server has caching of some kind.

For stream, if there was a way to ask the child server to paginate through the list, that might be an equivalent, but that is certainly not s always possible.

Fundamentally, however, I don’t see this case as much different that any field that has stream on it which is not backed by an async iterables. In all those cases, the server has all the results, but the client wants to receive them one by one… it’s a bit worse because the server has the full list plus alsi all the completed values and still has to wait to send, but if type merging is happening, the values may actually not be fully completed.

anyway, could be I didn’t 100% understand your comment above…

**Reply by michaelstaib** [reply]
I also think it should stay a should for the initial release. We are at the moment working to integrate it into our gateway and there are scenarios where it is better to not defer. We do not require the downstream services to support the directives but apply them in the gateway when we build the query plan. 

If the connection pattern is used we can even emulate stream since you essentially can fetch the initial count and then do a second call for the rest. Anyway, being able to skip defer and stream in some cases will improve fetching performance.

**Reply by yaacovCR** [reply]
I am not sure if we really have ever specified/quantified how exactly performance will be increased. It seems to me that what will change primarily is overall response length in that the initial payload will be sent separately, which can blow up the response into many payloads with deferred fields for lists.

We also may avoid a single page refresh although a debounce by client should handle this gracefully, and the deferred data may not even be currently visible/require rendering. Once batching is in place, we should avoid gui thrashing and the client will likely be doing denouncing.

Always deferring can potentially decrease the initial render time significantly if the deferred fields overall take a lot of time to transfer/decompress.

Overall, I am still strongly of the opinion that we should change language to MUST and perhaps add client specified option to allow the SHOULD allowable optimizations.

Do we have any real world numbers on this?

**Reply by michaelstaib** [reply]
The problem with the `allow optimizations` is that the client is at this point deployed and maybe does not receive further updates, but the server might change and the perf characteristics of the server might change so that at one point there might now be an optimization where defer makes no sense anymore.

**Reply by yaacovCR** [reply]
What I am suggesting is that this behavior SHOULD vs MUST is different enough to warrant supporting both. If you think clients would want SHOULD to allow optimizations then we should expect them to say so.

It sounds like you are saying that client developers may forget they can say so, and then regret it later when they realize they no longer need defer?

**Reply by michaelstaib** [reply]
No, I am saying it defeats the purpose if we introduced such a control.

**Reply by yaacovCR** [reply]
Could be. I am not understanding.

**Reply by yaacovCR** [reply]
> One reason why _should_ is better for advanced use cases (maybe not a _great_ reason but a reason):
> 
> If you have a server that composes many child servers (so each of those child servers' schemas compose to form the parent schema): some of those child servers may implement @defer/@stream, but if even one child server responsible for even a single schema type does not support it, you either need to walk the entire tree in the parent server in order to split up the already-stitched-together response tree (if all servers that support @defer/@stream _must_ batch every time), or we need to allow ignoring the defer/stream directives when the server chooses.

@mjmahone I just hit this example, and so understand your comment a bit better. I will flesh it out below with additional comments given the interim proposal changes.

Let's say I have a parent server with types:
```graphql
type Query {
  nodes: [SomeType]
}

type SomeType {
  someField: String
  anotherField: String
}
```

where fulfillment of the operation is via child servers A and B with types:

Child Server A:

```graphql
type SomeType {
  someField: String
}
```

and

Child Server B:

```graphql
type SomeType {
  anotherField: String
}
```

with query:
```graphql
query {
  nodes: {
    ... @defer {
      someField
      anotherField
    }
  }
}
```

What if child server A supports defer/stream, but child server B does not? Child server A may return:

```json
{
  "data": {
    "nodes": [
      N x
      {},
    ],
  },
}
```

```json
N x {
  "incremental":[
    {
      "path": ["nodes", n],
      "data": {
        "someField": "someData",
      } ,
    },
  ],
  "hasNext": usuallyTrueExceptLastN
}
```

and child server B may return:

```json
{
  "data": {
    "nodes": [
      N x
      {
        "anotherField": "otherData",
      },
    ],
  },
}
```

We want to enable a situation where we can just forward the merged results, rather than transforming results unnecessarily. If inlining is allowed, we can merge the initial results and send the following:

```json
{
  "data": {
    "nodes": [
      N x
      {
        "anotherField": "otherData",
      },
    ],
  },
}
```

```json
N x {
  "incremental":[
    {
      "path": ["nodes", n],
      "data": {
        "someField": "someData",
      } ,
    },
  ],
  "hasNext": usuallyTrueExceptLastN
}
```

But the above response would not be valid if we change SHOULD to MUST. If inlining is not allowed for the parent server, we have to "walk the initial result" from child server "B" and strip out the deferred bits and reconstruct the incremental response [^1][^2].
Note, that if inlining is allowed as according to the current proposal, then even if both child servers A and B support defer/stream, they could at any time decide to inline data, and the parent server would have to send the response-shape above, such that this would not be an uncommon scenario.

We do need to update this discussion in light of more recent discussions by the incremental delivery sub-WG. In particular, at recent meetings, it has been suggest that if defer/stream directives are to be ignored by advanced servers, they should either be honored or ignored for all of the fields within the fragment => that would render the above example response by the parent server invalid.

However, at even more recent meetings suggested that we allow incremental delivery of data in any format, but add `pending`/`completed` portions to the incremental map to indicate which paths are not complete. We could then augment the payload above with something TBD perhaps like:

```json
{
  "data": {
    "nodes": [
      N x
      {
        "anotherField": "otherData",
      },
    ],
  },
  "pending": [
    N x
    { "path": ["nodes", n] },
  ],
  ### or maybe a more concise one-liner:
  "pending": [{ "path": ["nodes", "*"] },],
  "hasNext": true,
}
```

```json
N x {
  "incremental": [
    {
      "path": ["nodes", n],
      "data": {
        "someField": "someData",
      } ,
      "completed": [{ "path": ["nodes", n] } ],
    },
  ],
  "hasNext": usuallyTrueExceptLastN
}
```

[^1]: Alternatively, what I suggested above is that we strip the deferred bits out from the get-go in terms of what we send to B, and get a second result in that fashion, hoping that the server appropriately used dataloaders so that the second response is not too delayed.
[^2]: Note that it's not exactly the "walking the initial result" that is the extra work, as we have to walk it anyway to perform the deep merge. It's the checking of the fields within that result as to whether they need to be deferred and then the creation of the objects representing the deferred patches that presumably does add some overhead.
